posthog/caching/warming.py

```diff
@@ -0,0 +1,139 @@
+from posthog.hogql_queries.legacy_compatibility.flagged_conversion_manager import conversion_to_query_based
+from posthog.hogql_queries.query_runner import ExecutionMode
+from posthog.models import Team, Insight, DashboardTile
+from posthog.schema import GenericCachedQueryResponse
+from posthog.tasks.utils import CeleryQueue
+
+logger = structlog.get_logger(__name__)
+
+STALE_INSIGHTS_COUNTER = Counter(
+    "posthog_cache_warming_stale_insights",
+    "Number of stale insights present",
+    ["team_id"],
+)
+PRIORITY_INSIGHTS_COUNTER = Counter(
+    "posthog_cache_warming_priority_insights",
+    "Number of priority insights warmed",
+    ["team_id", "dashboard", "is_cached"],
+)
+
+LAST_VIEWED_THRESHOLD = timedelta(days=7)
+
+
+def priority_insights(team: Team) -> Generator[tuple[int, Optional[int]], None, None]:
+    """
+    This is the place to decide which insights should be kept warm.
+    The reasoning is that this will be a yes or no decision. If we need to keep it warm, we try our best
+    to not let the cache go stale. There isn't any middle ground, like trying to refresh it once a day, since
+    that would be like clock that's only right twice a day.
+    """
+
+    threshold = datetime.now(UTC) - LAST_VIEWED_THRESHOLD
+    QueryCacheManager.cleanup_stale_insights(team_id=team.pk, threshold=threshold)
+    combos = QueryCacheManager.get_stale_insights(team_id=team.pk, limit=500)
+
+    STALE_INSIGHTS_COUNTER.labels(team_id=team.pk).inc(len(combos))
+
+    dashboard_q_filter = Q()
+    insight_ids_single = set()
+
+    for insight_id, dashboard_id in (combo.split(":") for combo in combos):
+        if dashboard_id:
+            dashboard_q_filter |= Q(insight_id=insight_id, dashboard_id=dashboard_id)
+        else:
+            insight_ids_single.add(insight_id)
+
+    if insight_ids_single:
+        single_insights = (
+            team.insight_set.filter(insightviewed__last_viewed_at__gte=threshold, pk__in=insight_ids_single)
+            .distinct()
+            .values_list("id", flat=True)
+        )
+        for single_insight_id in single_insights:
+            yield single_insight_id, None
+
+    if not dashboard_q_filter:
+        return
+
+    dashboard_tiles = (
+        DashboardTile.objects.filter(dashboard__last_accessed_at__gte=threshold)
+        .filter(dashboard_q_filter)
+        .distinct()
+        .values_list("insight_id", "dashboard_id")
+    )
+    yield from dashboard_tiles
+
+
+@shared_task(ignore_result=True, expires=60 * 60)
+def schedule_warming_for_teams_task():
+    team_ids = largest_teams()
+
+    logger.info("Warming insight cache: teams", team_ids=team_ids)
+
+    teams = Team.objects.filter(pk__in=team_ids)
+
+    if len(teams) == 0:  # For local development
+        teams = Team.objects.all()[:10]
+
+    for team in teams:
+        insight_tuples = priority_insights(team)
+
+        task_groups = chain(*(warm_insight_cache_task.si(*insight_tuple) for insight_tuple in insight_tuples))
```

Comment:
Loose question: how will we monitor that we are keeping up with these tasks? (I'm not too familiar with our Celery monitoring)